### ISSUE TITLE
How are _ready and  _process inherited?

### ISSUE THREAD

**Issue by aspenforest** [godotengine/godot-docs#2576]
I have a class which extends another class, which has _ready and _process functions.
If the child has also implemented _ready and _process, will they overwrite the parents implementation?
If the child has no _process of its own but inherits it from its parent, will calling self.set_process(true) enable the process which was inherited from the parent? 
Wish there was a decent documentation available about this.

**Comment by vortexofdoom** [comment]
I know they do not override the parent's implementations, it will call both, although I don't know in what order.

**Comment by aspenforest** [comment]
So what you are saying is that Godot doesn't really have actual inheritance.  
Thanks for the clarification.  
I'm really dissapointed that this essential information is not documented in an easily accessible manner.

**Comment by vortexofdoom** [comment]
I believe only certain functions like `_ready()`, `_process()`, and `_physics_process()` are affected by this. User-defined functions should be replaced.

It's possible to get around by calling an empty `_custom_function()` in the parent's `_process()` function and defining that behavior in the inheriting class. In my opinion this is sufficient, it's slightly hacky but only in theory, actual execution efficiency should be identical.

**Comment by Calinou** [comment]
You can call a parent function using the dot syntax: `.method()`

If the parent's virtual method isn't being run, you could likely run it this way.

**Comment by oldmerkum** [comment]
@aspenforest were you able to figure out the answer?
I can test this if not.

**Comment by aspenforest** [comment]
I haven't focused on this issue for a while, cause it hasn't been that
crucial and I have more pressing matters to attend to, but if you
would be able to come to a conclusion on this matter, then perhaps you
could make a pull request to the Godot Docs repo with the necessary
improvements?

Best regards,

**Comment by oldmerkum** [comment]
So, if the class extends Node, and during run-time is added as a child node, then the functions _ready(), process(delta), etc. are not overridden. Instead each implementation of the method is called.
If the class is not added as a child node and instead kept as a member variable, this is not the case.

Is this expected behaviour?

**Comment by oldmerkum** [comment]
Maybe this should be in the scripting section of the godot-docs?

**Comment by aspenforest** [comment]
"So, if the class extends Node, and during run-time is added as a child
node, then the functions _ready(), process(delta), etc. are not overridden."

Does it work the same way if one extends from an user-defined GDScript
file, which itself extends from e.g. Node?

If true, that is the biggest load of turd I have ever witnessed of ANY
programming language.

Thank you for your help on this matter, I appreciate it a lot. :)

**Comment by oldmerkum** [comment]
Yes, but again, that is because you are extending from a class that godot has implemented and most likely has some important instructions within its own _ready(), _process(delta), etc. methods.

So, I don't think it is load of turds, but necessary for the lazy game designer that does not want to implement their own Node class, nor care how it is implemented. You just want the functionality. Calling a custom function as @vortexofdoom described is a simple workaround.

In the _ready() function, call _custom_ready() which you defined. This _custom_ready() can then be overridden properly and still be called in _ready().

```gdscript
# base.gd
extends Node

class_name Base

func _ready():
  _whoami()

func _whoami():
  print("I am base")
```

```gdscript
# derived.gd
extends Base

class_name Derived

func _whoami():
  print("I am derived")
```
Also, if this is not in the docs, then it definitely should be. I will look and work on adding this information.

**Comment by aspenforest** [comment]
Great, thank you for the solution. :)
My resentment of this issue stemmed from the fact that Godot does not have
proper inheritance, but some mutant version of it, making things
unnecessarily complicated for devs who have background in real programming
langs like C++.

**Comment by oldmerkum** [comment]
Sure, I learned from this as well so it helped me too. Although, I sense some shade getting thrown at gdscript...

Can this be closed now?

**Comment by Calinou** [comment]
@oldmerkum Bump :slightly_smiling_face:

Are you still working on a pull request to add this information?

**Comment by oldmerkum** [comment]
I completely forgot that I offered to add it 😮
I will at least start a branch and figure out where to add that info so I don't forget again.
Thanks for the reminder.

**Comment by oldmerkum** [comment]
It looks like a recent commit added a short explanation of this:
https://github.com/godotengine/godot-docs/pull/2687/commits
> .. note:: Default functions like  ``_init``, and most notifications such as
>          ``_enter_tree``, ``_exit_tree``, ``_process``, ``_physics_process``,
>          etc. are called in all parent classes automatically.
>          There is no need to call them explicitly when overloading them.

Overriding functions was also already in the docs:
>  Functions override functions with the same name on the base/parent class.
>  If you still want to call them, use '.' (like 'super' in other languages).

The only other info that I could not find when first messing with Godot is that when you have to use the `add_child()` method for the built-in Godot functions to start automatically being called in an object that you create in a script. Which makes sense since it is not a part of the scene until it is added as a child, but took a day or two for me to figure that out.

Not sure if this topic deserves more notice in the docs; Seems like that sentence will be missed by skim readers like myself.

Finally, that merge I mentioned is only for the latest docs, not 3.1 stable, although it still applies to 3.1 stable. Can it be merged into the 3.1 branch as well?

**Comment by Calinou** [comment]
@oldmerkum I requested a cherry-pick into the `3.1` branch, we'll see if it's possible :slightly_smiling_face:
